Relaxed HomeBrew requirement

When I don't have HomeBrew, but dependencies are met, I'm still having problems running it due to a (non-critical?) dependency check:

```
2015-11-11 17:03:25 [ERROR] This script requires HomeBrew, but it's not installed. Aborting.
```

I'd propose relaxing the HomeBrew requirement and applying the check only in the install call.

--- a/src/docker-osx-dev
+++ b/src/docker-osx-dev
@@ -600,11 +600,6 @@ function check_prerequisites {
     log_error "This script should only be run on OS X"
     exit 1
   fi
-
-  if ! type brew > /dev/null 2>&1 ; then
-    log_error "This script requires HomeBrew, but it's not installed. Aborting."
-    exit 1
-  fi
 }
 
 #
@@ -627,16 +622,14 @@ function brew_install {
   local readonly command_name="$3"
   local readonly use_cask="$4"
 
-  local brew_command="brew"
-  if [[ "$use_cask" = true ]]; then
-    brew_command="brew cask"
-  fi
+  if ! type "$command_name" > /dev/null 2>&1 ; then
+    check_homebrew $readable_name
 
-  if eval "$brew_command list $package_name" > /dev/null 2>&1 ; then
-    log_warn "$readable_name is already installed by HomeBrew, skipping"
-  elif type "$command_name" > /dev/null 2>&1 ; then
-    log_warn "Found command $command_name, assuming $readable_name is already installed and skipping"
-  else
+    local brew_command="brew"
+    if [[ "$use_cask" = true ]]; then
+      brew_command="brew cask"
+    fi
+
     log_info "Installing $readable_name"
     eval "$brew_command install $package_name"
 
@@ -649,6 +642,20 @@ function brew_install {
       log_error "Failed to install dependency $readable_name. Check the log output above for reasons."
       exit 1
     fi
+  else
+    log_warn "Found command $command_name, assuming $readable_name is already installed and skipping"
+  fi
+}
+
+#
+# Checks if HomeBrew is installed.
+# Called only in case of a missing dependency.
+#
+function check_homebrew {
+  local readonly missing_package="$1"
+  if ! type brew > /dev/null 2>&1 ; then
+    log_error "HomeBrew is required to install $missing_package, but it's not installed. Aborting."
+    exit 1
   fi
 }
 

--- a/src/docker-osx-dev
+++ b/src/docker-osx-dev
@@ -630,8 +630,12 @@ function brew_install {
       brew_command="brew cask"
     fi
 
-    log_info "Installing $readable_name"
-    eval "$brew_command install $package_name"
+    if eval "$brew_command list $package_name" > /dev/null 2>&1 ; then
+      log_warn "$readable_name is already installed by HomeBrew, skipping"
+    else
+      log_info "Installing $readable_name"
+      eval "$brew_command install $package_name"
+    fi
 
     # It seems that brew install can fail to install a dependency, but exit
     # without an error code, so the set -e flag does not help us. Example:

--- a/src/docker-osx-dev
+++ b/src/docker-osx-dev
@@ -603,6 +603,20 @@ function check_prerequisites {
 }
 
 #
+# Update Brew formulae and HomeBrew
+#
+# Non-breaking. Falls through with warning if fails.
+#
+function brew_update {
+  log_info "Updating HomeBrew"
+  if ! type brew > /dev/null 2>&1 ; then
+    log_warn "HomeBrew not found, attempting to continue without"
+  else
+    brew update
+  fi
+}
+
+#
 # Usage: brew_install PACKAGE_NAME READABLE_NAME COMMAND_NAME USE_CASK
 #
 # Checks if PACKAGE_NAME is already installed by using brew as well as by
@@ -635,16 +649,16 @@ function brew_install {
     else
       log_info "Installing $readable_name"
       eval "$brew_command install $package_name"
-    fi
 
-    # It seems that brew install can fail to install a dependency, but exit
-    # without an error code, so the set -e flag does not help us. Example:
-    # https://github.com/brikis98/docker-osx-dev/issues/124
-    # Therefore, explicitly check that the dependency was installed before
-    # proceeding.
-    if ! type "$command_name" > /dev/null 2>&1; then
-      log_error "Failed to install dependency $readable_name. Check the log output above for reasons."
-      exit 1
+      # It seems that brew install can fail to install a dependency, but exit
+      # without an error code, so the set -e flag does not help us. Example:
+      # https://github.com/brikis98/docker-osx-dev/issues/124
+      # Therefore, explicitly check that the dependency was installed before
+      # proceeding.
+      if ! type "$command_name" > /dev/null 2>&1; then
+        log_error "Failed to install dependency $readable_name. Check the log output above for reasons."
+        exit 1
+      fi
     fi
   else
     log_warn "Found command $command_name, assuming $readable_name is already installed and skipping"
@@ -667,8 +681,7 @@ function check_homebrew {
 # Installs all the dependencies for docker-osx-dev.
 #
 function install_dependencies {
-  log_info "Updating HomeBrew"
-  brew update
+  brew_update
 
   brew_install "caskroom/cask/brew-cask" "Cask" "brew cask" false
   brew_install "boot2docker" "Boot2Docker" "boot2docker" false

